File: gaming_deals_bot/cheapshark.py
import logging
from dataclasses import dataclass

import httpx

logger = logging.getLogger(__name__)

BASE_URL = "https://www.cheapshark.com/api/1.0"

# CheapShark store IDs for PC/digital storefronts
STORE_IDS = {
    "1": "Steam",
    "7": "GOG",
    "11": "Humble Store",
    "23": "GreenManGaming",
}


@dataclass
class CheapSharkDeal:
    deal_id: str
    game_id: str
    title: str
    sale_price: str
    normal_price: str
    savings: float  # percentage off
    deal_rating: float
    store_id: str
    last_change: int  # unix timestamp
    steam_app_id: str | None

    @property
    def store_name(self) -> str:
        return STORE_IDS.get(self.store_id, f"Store {self.store_id}")

    @property
    def deal_url(self) -> str:
        return f"https://www.cheapshark.com/redirect?dealID={self.deal_id}"

    @property
    def dedup_id(self) -> str:
        return f"cheapshark-{self.game_id}-{self.last_change}"


async def fetch_deals(
    client: httpx.AsyncClient,
    *,
    max_price: float = 20,
    min_rating: float = 8.0,
    min_discount: float = 50,
    page_size: int = 10,
) -> list[CheapSharkDeal]:
    """Fetch top deals from CheapShark across configured stores."""
    store_ids = ",".join(STORE_IDS.keys())
    params = {
        "storeID": store_ids,
        "upperPrice": str(int(max_price)),
        "sortBy": "Deal Rating",
        "desc": "1",
        "pageSize": str(page_size),
    }

    try:
        resp = await client.get(f"{BASE_URL}/deals", params=params)
        resp.raise_for_status()
        raw_deals = resp.json()
    except (httpx.HTTPError, ValueError) as exc:
        logger.error("CheapShark API error: %s", exc)
        return []

    logger.info(
        "CheapShark returned %d raw deals (before filtering)", len(raw_deals),
    )

    deals = []
    for d in raw_deals:
        savings = float(d.get("savings", 0))
        rating = float(d.get("dealRating", 0))
        title = d.get("title", "?")

        if savings < min_discount:
            logger.debug("Filtered out %s: savings %.1f%% < %.1f%%", title, savings, min_discount)
            continue
        if rating > 0 and rating < min_rating:
            logger.debug("Filtered out %s: rating %.1f < %.1f", title, rating, min_rating)
            continue

        steam_app_id = d.get("steamAppID") or None
        if steam_app_id == "0":
            steam_app_id = None

        deals.append(
            CheapSharkDeal(
                deal_id=d["dealID"],
                game_id=d["gameID"],
                title=d["title"],
                sale_price=d["salePrice"],
                normal_price=d["normalPrice"],
                savings=savings,
                deal_rating=rating,
                store_id=d["storeID"],
                last_change=int(d.get("lastChange", 0)),
                steam_app_id=steam_app_id,
            )
        )

    logger.info("CheapShark returned %d deals after filtering", len(deals))
    return deals

File: internal/deals/cheapshark.go
package deals

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"math"
	"net/http"
	"strconv"
)

var storeNames = map[string]string{
	"1":  "Steam",
	"7":  "GOG",
	"11": "Humble Store",
	"23": "GreenManGaming",
}

type CheapSharkDeal struct {
	DealID       string
	GameID       string
	Title        string
	SalePrice    float64
	NormalPrice  float64
	Savings      float64
	DealRating   float64
	StoreID      string
	StoreName    string
	LastChange   int64
	SteamAppID   string
	DealURL      string
	IsHistLow    bool
	DedupID      string
}

type cheapSharkRaw struct {
	DealID     string `json:"dealID"`
	GameID     string `json:"gameID"`
	Title      string `json:"title"`
	SalePrice  string `json:"salePrice"`
	NormalPrice string `json:"normalPrice"`
	Savings    string `json:"savings"`
	DealRating string `json:"dealRating"`
	StoreID    string `json:"storeID"`
	LastChange int64  `json:"lastChange"`
	SteamAppID string `json:"steamAppID"`
}

// FetchCheapSharkDeals fetches deals from CheapShark API.
func FetchCheapSharkDeals(maxPrice float64, pageSize int) ([]CheapSharkDeal, error) {
	url := fmt.Sprintf(
		"https://www.cheapshark.com/api/1.0/deals?storeID=1,7,11,23&upperPrice=%d&sortBy=recent&desc=1&pageSize=%d",
		int(maxPrice), pageSize,
	)

	resp, err := http.Get(url)
	if err != nil {
		return nil, fmt.Errorf("cheapshark request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("cheapshark returned status %d", resp.StatusCode)
	}

	var raw []cheapSharkRaw
	if err := json.NewDecoder(resp.Body).Decode(&raw); err != nil {
		return nil, fmt.Errorf("cheapshark decode failed: %w", err)
	}

	var deals []CheapSharkDeal
	for _, d := range raw {
		sale, _ := strconv.ParseFloat(d.SalePrice, 64)
		normal, _ := strconv.ParseFloat(d.NormalPrice, 64)
		savings, _ := strconv.ParseFloat(d.Savings, 64)
		rating, _ := strconv.ParseFloat(d.DealRating, 64)

		storeName := storeNames[d.StoreID]
		if storeName == "" {
			storeName = "Unknown"
		}

		deals = append(deals, CheapSharkDeal{
			DealID:      d.DealID,
			GameID:      d.GameID,
			Title:       d.Title,
			SalePrice:   sale,
			NormalPrice:  normal,
			Savings:     savings,
			DealRating:  rating,
			StoreID:     d.StoreID,
			StoreName:   storeName,
			LastChange:  d.LastChange,
			SteamAppID:  d.SteamAppID,
			DealURL:     fmt.Sprintf("https://www.cheapshark.com/redirect?dealID=%s", d.DealID),
			DedupID:     fmt.Sprintf("cheapshark-%s-%d", d.GameID, d.LastChange),
		})
	}

	return deals, nil
}

// FilterCheapSharkDeals filters deals by rating, discount, and price thresholds.
func FilterCheapSharkDeals(deals []CheapSharkDeal, minRating float64, minDiscount int, maxPrice float64) []CheapSharkDeal {
	var filtered []CheapSharkDeal
	for _, d := range deals {
		discount := int(math.Floor(d.Savings))
		if discount < minDiscount {
			slog.Debug("cheapshark: skipping (low discount)", "title", d.Title, "discount", discount)
			continue
		}
		if d.DealRating < minRating && d.DealRating != 0 {
			slog.Debug("cheapshark: skipping (low rating)", "title", d.Title, "rating", d.DealRating)
			continue
		}
		if d.SalePrice > maxPrice {
			slog.Debug("cheapshark: skipping (too expensive)", "title", d.Title, "price", d.SalePrice)
			continue
		}
		filtered = append(filtered, d)
	}
	return filtered
}
